yellow-paper/docs/contract-deployment/classes.md

```diff
@@ -59,26 +86,70 @@ Note the lack of visibility modifiers. Internal functions are specified as a mac
 
 Also note the lack of commitment to the function compilation artifact. Even though a commitment to a function is required so that the PXE can verify the execution of correct unconstrained Brillig code embedded within private functions, this is handled entirely out of protocol. As such, PXEs are expected to verify it against the `artifact_hash` in the containing contract class.
 
+#### Private Function Leaf Hash
+
+<!-- HASH DEFINITION -->
+
+```rust
+private_function_leaf_crh(
+    f: PrivateFunction
+) -> Field {
+    let private_function_leaf = poseidon2(
+        be_string_to_field("az_private_function_leaf"),
+
+        be_bits_to_field(f.function_selector),
+        f.vk_hash
+    );
+
+    private_function_leaf
+}
+```
+
 ### Artifact Hash
 
 Even though not enforced by the protocol, it is suggested for the `artifact_hash` to follow this general structure, in order to be compatible with the definition of the [`broadcast` function below](#broadcast).
 
-```
-private_functions_artifact_leaves = artifact.private_functions.map(fn =>
-  sha256(fn.selector, fn.metadata_hash, sha256(fn.private_bytecode))
-)
-private_functions_artifact_tree_root = merkleize(private_functions_artifact_leaves)
+<!-- HASH DEFINITION -->
 
-unconstrained_functions_artifact_leaves = artifact.unconstrained_functions.map(fn =>
-  sha256(fn.selector, fn.metadata_hash, sha256(fn.unconstrained_bytecode))
-)
-unconstrained_functions_artifact_tree_root = merkleize(unconstrained_functions_artifact_leaves)
+```rust
+artifact_crh(
+  artifact // This type is out of protocol, e.g. the format output by Nargo
+) -> Field {
 
-artifact_hash = sha256(
-  private_functions_artifact_tree_root,
-  unconstrained_functions_artifact_tree_root,
-  artifact_metadata_hash,
-)
+  let private_functions_artifact_leaves: [u8; 32][] = artifact.private_functions.map(|f|
+    sha256(
+      be_string_to_bits("az_artifact_private_function_leaf"),
+
+      f.selector, // 32-bits
+      f.metadata_hash, // 256-bits
+      sha256(f.private_bytecode)
+    )
+  );
+  let private_functions_artifact_tree_root: [u8; 32] = merkleize(private_functions_artifact_leaves);
+
+  let unconstrained_functions_artifact_leaves: [u8; 32][] = artifact.unconstrained_functions.map(|f|
+    sha256(
+      be_string_to_bits("az_artifact_unconstrained_function_leaf"),
+
+      f.selector, // 32-bits
+      f.metadata_hash, // 256-bits
+      sha256(f.unconstrained_bytecode)
+    )
+  );
+  let unconstrained_functions_artifact_tree_root: [u8; 32] = merkleize(unconstrained_functions_artifact_leaves);
+
+  let artifact_hash_256_bit: [u8; 32] = sha256(
+    be_string_to_field("az_artifact"),
+
+    private_functions_artifact_tree_root, // 256-bits
+    unconstrained_functions_artifact_tree_root, // 256-bits
+    artifact_metadata
+  );
+
+  let artifact_hash: Field = artifact_hash_256_bit % FIELD_MODULUS;
```

Comment:
Should we just update all these hashes to be poseidon, for consistency? sha256 would make sense for simplicity and speed, but it seems quite ugly that we're then doing field operations on all the values anyway (modulo), and padding all the input values from fields to 256-bits.
It's not an urgently-needed change (if you were to agree), since you've already implemented it with sha256.

Reply:
I don't know how faster is sha256 than poseidon to make that call